liquid-rust-typeck/src/global_env.rs

```diff
@@ -1,27 +1,60 @@
+use std::cell::RefCell;
+
+use liquid_rust_common::errors::ErrorReported;
+use liquid_rust_core::resolve::Resolver;
 use rustc_hash::FxHashMap;
 use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_middle::ty::TyCtxt;
 pub use rustc_middle::ty::Variance;
+pub use rustc_span::symbol::Ident;
+use rustc_span::Span;
 
 use crate::ty::{self, BaseTy, Sort};
 
 pub struct GlobalEnv<'tcx> {
-    pub fn_specs: FxHashMap<LocalDefId, ty::FnSpec>,
+    pub fn_specs: RefCell<FxHashMap<LocalDefId, ty::FnSpec>>,
     pub adt_defs: FxHashMap<LocalDefId, ty::AdtDef>,
     pub tcx: TyCtxt<'tcx>,
 }
 
 impl<'tcx> GlobalEnv<'tcx> {
     pub fn new(
         tcx: TyCtxt<'tcx>,
-        fn_specs: FxHashMap<LocalDefId, ty::FnSpec>,
+        fn_specs: RefCell<FxHashMap<LocalDefId, ty::FnSpec>>,
         adt_defs: FxHashMap<LocalDefId, ty::AdtDef>,
     ) -> Self {
         GlobalEnv { fn_specs, adt_defs, tcx }
     }
 
-    pub fn lookup_fn_sig(&self, did: DefId) -> &ty::Binders<ty::FnSig> {
-        &self.fn_specs[&did.as_local().unwrap()].fn_sig
+    fn lookup_default_spec(
+        &self,
+        local_def: LocalDefId,
+        span: Span,
+    ) -> Result<ty::FnSpec, ErrorReported> {
+        let mut resolver = Resolver::from_fn(self.tcx, local_def)?;
+        let fn_sig = resolver.default_sig(local_def, span)?;
+        // TODO -- wf needs a bunch of stuff not in scope here.
+        // of course "default" sigs SHOULD be well-formed so perhaps redundant
+        // wf.check_fn_sig(&fn_sig)?;
```

Comment:
I like these redundant checks. We should at least enable them in debug builds and do as many as we can.